### PR TITLE
--validate: Check server version for minimum version number.

### PR DESCRIPTION
**Pull request by G-Harmon**
We need a minimum of v1.8.1 for MCI to work.

This is for Issue #53 .

cc @nikhiljindal

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/167)

### CODE REVIEW

File: app/kubemci/cmd/create.go [hunk 1]
```diff
@@ -149,6 +149,14 @@ func runCreate(options *CreateOptions, args []string) error {
 		return err
 	}
 
+	if options.Validate {
+		if newEnough, err := validations.ServerVersionsNewEnough(clients); err != nil {
+			return fmt.Errorf("cluster version could not be validated: %s", err)
+		} else if !newEnough {
+			return fmt.Errorf("not all cluster versions new enough for kubemci")
```
Comment (nikhiljindal) [inline app/kubemci/cmd/create.go]: We should be explicit about what "not new enough" means. Say they should be > 1.8.1 and not 1.10.0?

File: app/kubemci/pkg/validations/validations.go [hunk 2]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
 	}
 	return nil
 }
+
+func ServerVersionsNewEnough(clients map[string]kubeclient.Interface) (bool, error) {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: Does this have to be public?
Call this from the existing public Validate method?

File: app/kubemci/pkg/validations/validations.go [hunk 3]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+	for key := range clients {
+		glog.Infof("Checking client %s", key)
+		discoveryClient := clients[key].Discovery()
+		if discoveryClient == nil {
+			return false, fmt.Errorf("no discovery client in client: %+v", clients[key])
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: Also print the cluster name (key).

File: app/kubemci/pkg/validations/validations.go [hunk 4]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+			fmt.Println("Cluster", key, "is not running a supported kubernetes version. Need at least 1.8.1.")
+			return false, nil
+		}
+
+	}
+	return true, nil
+}
+
+func serverVersionNewEnough(version string) (bool, error) {
+	// Example string we're matching: v1.9.3-gke.0
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: FYI thats for GKE clusters. Non GKE clusters will return "v1.9.3", without the gke suffix. We should test both in unit test

File: app/kubemci/pkg/validations/validations.go [hunk 5]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+	// Patch version
+	patch, err := strconv.ParseUint(matches[3], 10 /*bases*/, 32 /*bitSize*/)
+	if err != nil {
+		return false, err
+	}
+	glog.V(2).Infof("Got version: major:", major, "minor:", minor, "patch:", patch)
+	if patch >= 1 {
+		return true, nil
+	}
+	return false, nil
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: Also check !1.10.0? Or you want to do that in a separate PR?

File: app/kubemci/pkg/validations/validations.go [hunk 6]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+	major, err := strconv.ParseUint(matches[1], 10 /*base*/, 32 /*bitSize*/)
+	if err != nil {
+		return false, err
+	}
+	if major > 1 {
+		return true, nil
+	} else if major < 1 {
+		return false, nil
+	}
+	//Minor version
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: nit: add a space after "//"

File: app/kubemci/pkg/validations/validations.go [hunk 7]
```diff
@@ -89,3 +91,69 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+		if !newEnough {
+			fmt.Println("Cluster", key, "is not running a supported kubernetes version. Need at least 1.8.1.")
+			return false, nil
+		}
+
+	}
+	return true, nil
+}
+
+func serverVersionNewEnough(version string) (bool, error) {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: Add a unit test for this.
Recommend splitting this into
parseServerVersion() and validateServerVersion() and then you can test both independently.

File: app/kubemci/pkg/validations/validations.go (hunk 2 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: I think it needs to be public. validations.Validate is called from loadbalancersyncer.go. ServerVersionsNewEnough is called from create.go. With this setup, we check cluster versions before calling EnsureIngress().

File: app/kubemci/pkg/validations/validations.go (hunk 3 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: done.

File: app/kubemci/pkg/validations/validations.go (hunk 4 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: most of the tests are without the gke suffix, but i've now added one with "-gke.0"

File: app/kubemci/pkg/validations/validations.go (hunk 6 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: done.

File: app/kubemci/cmd/create.go (hunk 1 above)
Comment (G-Harmon) [reply app/kubemci/cmd/create.go]: I wanted to keep the knowledge of the exact supported versions in a minimum of locations. I've centralized it all in validations.go. That prints
fmt.Println("Cluster", key, "is not running a supported kubernetes version. Need at least 1.8.1.")
I'll the testing and logging about 1.10.0.

File: app/kubemci/pkg/validations/validations.go (hunk 7 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: oops. I wrote a test in validations_test.go, but somehow it got lost before uploading this PR.

File: app/kubemci/pkg/validations/validations.go (hunk 5 above)
Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: Done. And it was much easier after factoring it out into a "parse version number" function.

File: app/kubemci/pkg/validations/validations_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		in    string
+		isErr bool
+		major uint64
+		minor uint64
+		patch uint64
+	}{
+		{"v1.2.3", false, 1, 2, 3},
+		{"v10.200.3000", false, 10, 200, 3000},
+		{"v0.8.a", true, 0, 0, 0},
+		{"v1.-2.3", true, 1, 2, 3},
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations_test.go]: Add v1.9.3-gke.0 as a test case. Also one without "v". Like 1.2.3

File: app/kubemci/pkg/validations/validations_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+		{2, 0, 0, true},
+		// Test minor
+		{1, 7, 0, false},
+		{1, 7, 14, false},
+		{1, 9, 0, true},
+		// Test patch.
+		{1, 8, 0, false},
+		{1, 8, 1, true},
+
+		// 1.10.0 was buggy and doesn't work.
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations_test.go]: optional: You can link to https://github.com/kubernetes/ingress-gce/issues/182 in case reader is curious

File: app/kubemci/pkg/validations/validations_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+			t.Errorf("ServerVerNewEnough(%d, %d, %d): Expected no err. Got err %v", tt.major, tt.minor, tt.patch, err)
+		}
+		if tt.newEnough != ok {
+			t.Errorf("ServerVerNewEnough(%d, %d, %d): Expected newEnough? %v. Got newEnough:%v", tt.major, tt.minor, tt.patch, tt.newEnough, ok)
+		}
+	}
+}
+
+func TestVersionsAcrossClusters(t *testing.T) {
+
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations_test.go]: nit: line break not needed

File: app/kubemci/pkg/validations/validations.go [hunk 11]
```diff
@@ -89,3 +91,84 @@ func nodePortSameInAllClusters(backend v1beta1.IngressBackend, namespace string,
...
+			fmt.Printf("Cluster %s (ver %d.%d.%d) is not running a supported kubernetes version. Need >= 1.8.1 and not 1.10.0.\n",
+				key, major, minor, patch)
+			return false, nil
+		}
+
+	}
+	return true, nil
+}
+
+func serverVersionNewEnough(major, minor, patch uint64) (bool, error) {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/validations/validations.go]: This never returns an error so no need to include it in method signatures

File: app/kubemci/cmd/create.go (hunk 1 above)
Comment (nikhiljindal) [reply app/kubemci/cmd/create.go]: hmm. Seems like you can just remove the bool here. If err !=nil, then return that as is.
ServerVersionsNewEnough already returns meaningful error messages.

File: app/kubemci/pkg/validations/validations.go (hunk 2 above)
Comment (nikhiljindal) [reply app/kubemci/pkg/validations/validations.go]: We should do complete validation first. So validations.Validate() should also be called from create.go before calling EnsureIngress()?

Comment (G-Harmon) [reply app/kubemci/pkg/validations/validations.go]: It looks like it works okay to move all the validation code into Validate and calling it earlier in create.go. I'm having a little trouble moving the unit tests over cleanly. (will followup on the unit tests...)

**Comment by nikhiljindal** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[app/kubemci/pkg/validations/validations_test.go, line 38 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/167#-L9E0HiWRHvyz4uDqrkN-r1-38:-L9Eb5iwB_PdgRtT5PnU:b-1gtqli) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/df669b490331cc5a6d870ca51a5f2f03259719c0/app/kubemci/pkg/validations/validations_test.go#L38)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Are you sure it's valid to not start with "v"? Right now, I'm including that in my regular expression: ^v([0-9]*).([0-9]*).([0-9]*)

Added gke version string.
</blockquote></details>


invalid is fine. i was pointing out that you dont have any case without the v prefix

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/167)*

**Comment by nikhiljindal** [comment]
/lgtm
thanks!
